File: app/src/main/java/com/kotlinchina/smallpockets/widget/CusTomProgressWebView.java
package com.kotlinchina.smallpockets.widget;

import android.content.Context;
import android.graphics.Bitmap;
import android.util.AttributeSet;
import android.view.KeyEvent;
import android.view.View;
import android.webkit.JsPromptResult;
import android.webkit.JsResult;
import android.webkit.WebChromeClient;
import android.webkit.WebSettings;
import android.webkit.WebView;
import android.webkit.WebViewClient;
import android.widget.LinearLayout;
import android.widget.ProgressBar;

import com.kotlinchina.smallpockets.R;

/**
 *
 */
public class CusTomProgressWebView extends LinearLayout {

	private WebView mWebView;

	private ProgressBar mProgressBar;

	private Context mContext;

	private String url;

//	private String errorHtml = "<html><head><meta charset='UTF-8'></head><body><br><br><br><br><br><br><br><div align='center' style='font-size: smaller'  onclick='window.android.refresh()' ><a href='http://www.baidu.com' style='text-decoration: none'>暂无数据 <br/> 点击调用android方法 </a></div></body></html>";

//	@JavascriptInterface
//	public void refresh() {
//		Toast.makeText(mContext, "js 调用方法", Toast.LENGTH_SHORT).show();
//	}


	public CusTomProgressWebView(Context context) {
		this(context, null);
	}


	public CusTomProgressWebView(Context context, AttributeSet attrs) {
		this(context, attrs, 0);
	}

	public CusTomProgressWebView(Context context, AttributeSet attrs, int defStyle) {
		super(context, attrs, defStyle);
		this.mContext = context;
		initView(context);
	}

	private void initView(Context context) {
		View view = View.inflate(context, R.layout.view_web_progress, this);
		mWebView = (WebView) view.findViewById(R.id.web_view) ;
		mProgressBar = (ProgressBar) view.findViewById(R.id.progress_bar) ;
	}

	public String getUrl() {
		return url;
	}

	public void setUrl(String url) {
		this.url = url;
	}

	public void loadUrl(String url) {
		if(url !=null)
		initWebview(url);
	}


	private void initWebview(String url) {

		mWebView.addJavascriptInterface(this, "android");

		WebSettings webSettings = mWebView.getSettings();

		webSettings.setJavaScriptEnabled(true);
		// 设置可以访问文件
		webSettings.setAllowFileAccess(true);
		// 设置可以支持缩放
		webSettings.setSupportZoom(true);
		// 设置默认缩放方式尺寸是far
		webSettings.setDefaultZoom(WebSettings.ZoomDensity.MEDIUM);
		// 设置出现缩放工具
		webSettings.setBuiltInZoomControls(false);
		webSettings.setDefaultFontSize(16);

		mWebView.loadUrl(url);

		// 设置WebViewClient
		mWebView.setWebViewClient(new WebViewClient() {
			// url拦截
			@Override
			public boolean shouldOverrideUrlLoading(WebView view, String url) {
				// 使用自己的WebView组件来响应Url加载事件，而不是使用默认浏览器器加载页面
				view.loadUrl(url);
				// 相应完成返回true
				return true;
				// return super.shouldOverrideUrlLoading(view, url);
			}

			// 页面开始加载
			@Override
			public void onPageStarted(WebView view, String url, Bitmap favicon) {
				mProgressBar.setVisibility(View.VISIBLE);
				super.onPageStarted(view, url, favicon);
			}

			// 页面加载完成
			@Override
			public void onPageFinished(WebView view, String url) {
				mProgressBar.setVisibility(View.GONE);
				super.onPageFinished(view, url);
			}

			// WebView加载的所有资源url
			@Override
			public void onLoadResource(WebView view, String url) {
				super.onLoadResource(view, url);
			}

			@Override
			public void onReceivedError(WebView view, int errorCode, String description, String failingUrl) {
//				view.loadData(errorHtml, "text/html; charset=UTF-8", null);
				super.onReceivedError(view, errorCode, description, failingUrl);
			}

		});

		// 设置WebChromeClient
		mWebView.setWebChromeClient(new WebChromeClient() {
			@Override
			// 处理javascript中的alert
			public boolean onJsAlert(WebView view, String url, String message, final JsResult result) {
				return super.onJsAlert(view, url, message, result);
			};

			@Override
			// 处理javascript中的confirm
			public boolean onJsConfirm(WebView view, String url, String message, final JsResult result) {
				return super.onJsConfirm(view, url, message, result);
			};

			@Override
			// 处理javascript中的prompt
			public boolean onJsPrompt(WebView view, String url, String message, String defaultValue, final JsPromptResult result) {
				return super.onJsPrompt(view, url, message, defaultValue, result);
			};

			// 设置网页加载的进度条
			@Override
			public void onProgressChanged(WebView view, int newProgress) {
				mProgressBar.setProgress(newProgress);
				super.onProgressChanged(view, newProgress);
			}

			// 设置程序的Title
			@Override
			public void onReceivedTitle(WebView view, String title) {
				super.onReceivedTitle(view, title);
			}
		});
		mWebView.setOnKeyListener(new OnKeyListener() {
			@Override
			public boolean onKey(View v, int keyCode, KeyEvent event) {
				if (event.getAction() == KeyEvent.ACTION_DOWN) {
					if (keyCode == KeyEvent.KEYCODE_BACK && mWebView.canGoBack()) { // 表示按返回键

						mWebView.goBack(); // 后退

						// webview.goForward();//前进
						return true; // 已处理
					}
				}
				return false;
			}
		});
	}

	public boolean canBack() {
		if (mWebView.canGoBack()) {
			mWebView.goBack();
			return false;
		}
		return true;
	}
}

File: app/src/main/java/com/kotlinchina/smallpockets/widget/CusTomProgressWebView.kt
package com.kotlinchina.smallpockets.widget

import android.content.Context
import android.graphics.Bitmap
import android.util.AttributeSet
import android.view.KeyEvent
import android.view.View
import android.webkit.JsPromptResult
import android.webkit.JsResult
import android.webkit.WebChromeClient
import android.webkit.WebSettings
import android.webkit.WebView
import android.webkit.WebViewClient
import android.widget.LinearLayout
import android.widget.ProgressBar
/**
 * Created by junjun
 */
import com.kotlinchina.smallpockets.R

class CustomProgressWebView : LinearLayout {

    private var mWebView: WebView? = null
    private var mProgressBar: ProgressBar? = null
    constructor(context: Context?) : this(context,null){}
    constructor(context: Context?, attrs: AttributeSet?) : this(context, attrs,0) {}
    constructor(context: Context?, attrs: AttributeSet?, defStyle: Int) : super(context, attrs, defStyle) {
        initView(context)
    }

    private fun initView(context: Context?) {
        val view = View.inflate(context, R.layout.view_web_progress, this)
        mWebView = view?.findViewById(R.id.web_view) as? WebView
        mProgressBar = view?.findViewById(R.id.progress_bar) as? ProgressBar
    }

    fun loadUrl(url: String?) {
        var url = url
        if (url == null) {
            url = "http://www.baidu.com"
        }
        initWebview(url)
    }

    private fun initWebview(url: String) {
        initBaseSetting(url)
        setWebViewClient()
        setWebchromeClient()
        webViewOnkeyListener()
    }

    private fun initBaseSetting(url: String) {
        mWebView?.addJavascriptInterface(this, "android")
        val webSettings = mWebView?.settings
        webSettings?.javaScriptEnabled = true
        webSettings?.allowFileAccess = true
        webSettings?.setSupportZoom(true)
        webSettings?.defaultZoom = WebSettings.ZoomDensity.MEDIUM
        webSettings?.builtInZoomControls = false
        webSettings?.defaultFontSize = 16
        mWebView?.loadUrl(url)
    }

    private  fun setWebViewClient(){
        mWebView?.setWebViewClient(object : WebViewClient() {
            override fun shouldOverrideUrlLoading(view: WebView?, url: String): Boolean {
                view?.loadUrl(url)
                return true
            }

            override fun onPageStarted(view: WebView?, url: String?, favicon: Bitmap?) {
                mProgressBar?.visibility = View.VISIBLE
                super.onPageStarted(view, url, favicon)
            }

            override fun onPageFinished(view: WebView?, url: String?) {
                mProgressBar?.visibility = View.GONE
                super.onPageFinished(view, url)
            }

            override fun onLoadResource(view: WebView?, url: String?) {
                super.onLoadResource(view, url)
            }

            override fun onReceivedError(view: WebView?, errorCode: Int, description: String, failingUrl: String) {
                super.onReceivedError(view, errorCode, description, failingUrl)
            }

        })
    }

    private fun setWebchromeClient() {
        mWebView?.setWebChromeClient(object : WebChromeClient() {
            override
            fun onJsAlert(view: WebView, url: String, message: String, result: JsResult): Boolean {
                return super.onJsAlert(view, url, message, result)
            }

            override
            fun onJsConfirm(view: WebView, url: String, message: String, result: JsResult): Boolean {
                return super.onJsConfirm(view, url, message, result)
            }

            override
            fun onJsPrompt(view: WebView, url: String, message: String, defaultValue: String, result: JsPromptResult): Boolean {
                return super.onJsPrompt(view, url, message, defaultValue, result)
            }

            override fun onProgressChanged(view: WebView, newProgress: Int) {
                mProgressBar?.progress = newProgress
                super.onProgressChanged(view, newProgress)
            }

            override fun onReceivedTitle(view: WebView, title: String) {
                super.onReceivedTitle(view, title)
            }
        })
    }

    private fun webViewOnkeyListener() {
        mWebView?.setOnKeyListener(View.OnKeyListener { v, keyCode, event ->
            if (event.action == KeyEvent.ACTION_DOWN) {
                val canGoBack = mWebView?.canGoBack() as? Boolean
                if (keyCode == KeyEvent.KEYCODE_BACK && canGoBack!=null && canGoBack) {
                    mWebView?.goBack()
                    return@OnKeyListener true
                }
            }
            false
        })
    }

}
